src/solc_0.8/defi/rules/ContributionRules.sol

```diff
@@ -0,0 +1,194 @@
+
+    struct MultiplierERC721 {
+        uint256[] ids;
+        uint256[] multiplier;
+        uint256 index;
+        bool balanceOf;
+    }
+
+    struct MultiplierERC1155 {
+        IERC1155 assetContract;
+        uint256[] ids;
+        uint256[] multiplier;
+        uint256 index;
+    }
+
+    mapping(IERC721 => MultiplierERC721) private _listERC721;
+    mapping(IERC1155 => MultiplierERC1155) private _listERC1155;
+    IERC721[] internal _listERC721Index;
+    IERC1155[] internal _listERC1155Index;
+
+    IERC721 public landMultiplierContract;
+
+    constructor(IERC721 _landMultiplierContract) {
+        landMultiplierContract = _landMultiplierContract;
+    }
+
+    function computeMultiplier(address account, uint256 amountStaked) external view returns (uint256) {
+        uint256 nftMultiplier = _multiplierBalanceOfERC721(account);
+        uint256 assetMultiplier = _multiplierBalanceOfERC1155(account);
+        uint256 landMultiplier = _multiplierBalanceOfLand(account);
+
+        return amountStaked + ((amountStaked * (nftMultiplier + assetMultiplier + landMultiplier)) / 100);
+    }
+
+    function setLandMultiplierContract(address newLandContract) external onlyOwner {
+        require(newLandContract.isContract(), "ContributionRules: invalid contract address");
+        landMultiplierContract = IERC721(newLandContract);
+    }
+
+    function addERC1155MultiplierList(
+        address contractERC1155,
+        uint256[] memory ids,
+        uint256[] memory multiplier
+    ) external onlyOwner {
+        require(contractERC1155.isContract(), "ContributionRules: invalid contract address");
+        require(ids.length > 0, "ContributionRules: invalid ids array");
+        require(multiplier.length > 0, "ContributionRules: invalid multiplier array");
+
+        IERC1155 multContract = IERC1155(contractERC1155);
+
+        _listERC1155[multContract].ids = ids;
+        _listERC1155[multContract].multiplier = multiplier;
+
+        if (_islistERC1155Member(multContract) == false) {
+            _listERC1155Index.push(multContract);
+            _listERC1155[multContract].index = _listERC1155Index.length - 1;
+        }
+    }
+
+    function addERC721MultiplierList(
+        address contractERC721,
+        bool balanceOf,
+        uint256[] memory ids
+    ) external onlyOwner {
+        require(contractERC721.isContract(), "ContributionRules: invalid contract address");
+        require(
+            (balanceOf == true && ids.length == 0) || (balanceOf == false && ids.length > 0),
+            "ContributionRules: invalid ids array"
+        );
+
+        IERC721 multContract = IERC721(contractERC721);
+
+        if (balanceOf == false) {
+            _listERC721[multContract].ids = ids;
+        }
+
+        if (_islistERC721Member(multContract) == false) {
+            _listERC721Index.push(multContract);
+            _listERC721[multContract].index = _listERC1155Index.length - 1;
+        }
```

Comment:
else revert ?

Reply:
we gonna keep it as it will be used to add and update